### ISSUE TITLE
Destructuring undefined function argument potential bug

### ISSUE THREAD

**Issue by joshribakoff** [microsoft/TypeScript#40034]
**TypeScript Version:** 
3.8.3 (also same on 3.9.7)

**Search Terms:**
Destructuring undefined function argument

**Code**
https://github.com/rx-store/rx-store/pull/48/commits/d52e626c80b2eb299accfef3ecb056ec6765ae90#diff-95f7c390f2d0664110959b4242d0ce4dR27-R48

**Expected behavior:**
Destructuring arg should yield the same types regardless of if its done in the function signature or the function body

**Actual behavior:**
The type loses the `|undefined` if its not destructured with extra indirection:

Example of unexpected results, `|undefined` is missing:

<img width="701" alt="Screen Shot 2020-08-13 at 6 55 52 AM" src="https://user-images.githubusercontent.com/4021306/90143510-8e13ff80-dd32-11ea-9c55-307750f78c66.png">

But here, if simply moved to another line (the function body), it does have the correct type:
<img width="926" alt="Screen Shot 2020-08-13 at 6 55 31 AM" src="https://user-images.githubusercontent.com/4021306/90143512-8fddc300-dd32-11ea-9505-217f9651df77.png">

<img width="677" alt="Screen Shot 2020-08-13 at 6 55 17 AM" src="https://user-images.githubusercontent.com/4021306/90143515-90765980-dd32-11ea-9ccb-85e01b694551.png">

**Playground Link:** 
https://github.com/rx-store/rx-store/pull/48/commits/d52e626c80b2eb299accfef3ecb056ec6765ae90#diff-95f7c390f2d0664110959b4242d0ce4dR27-R48

**Comment by joshribakoff** [comment]
Actually not even the workaround works :(

<img width="497" alt="Screen Shot 2020-08-13 at 7 06 12 AM" src="https://user-images.githubusercontent.com/4021306/90144293-8012ae80-dd33-11ea-8152-7f6d13ed9e70.png">
<img width="493" alt="Screen Shot 2020-08-13 at 7 06 06 AM" src="https://user-images.githubusercontent.com/4021306/90144295-80ab4500-dd33-11ea-9553-1c35136ece90.png">

This has got to be a bug right?

**Comment by joshribakoff** [comment]
Maybe the same problem, this time no destructuring...

<img width="751" alt="Screen Shot 2020-08-13 at 7 38 14 AM" src="https://user-images.githubusercontent.com/4021306/90148235-f74a4180-dd37-11ea-801d-6c64e3b60c75.png">

Seems to happen even here:

```
function test(foo?: {bar:'string'}|undefined) {
  console.log(foo.bar)
}
```

Turning strict mode back ON fixed it (I thought we had it on already). I'm not sure why anyone would want the runtime error, but feel free to close this if its expected, I guess.

**Comment by DanielRosenwasser** [comment]
I'm just not seeing this in the playground:

[Playground Link](https://www.typescriptlang.org/play?strictNullChecks=false&ts=4.1.0-dev.20200807#code/GYVwdgxgLglg9mABFApgZygCmHOB+ALkQG8AjAQwCcCByDSmMAcxoF8AfcAExWEZS4BKEgChEiCAjRwANigB0MuE2y55FSoJGsgA)

**Comment by joshribakoff** [comment]
Same, that’s what hinted to me that I needed to maybe enable strict mode to get the error or it was due to some other thing in our project setup. Our repo was recently overhauled with Nx.dev so it’s using a pretty nuanced project references with numerous TS configs all over the place.

I’m unblocked, feel free to close this or let me know if I can provide any more info to be of help

**Comment by RyanCavanaugh** [comment]
Closing; feel free to open a new issue if you have concrete repro steps. Thanks!
